go/store/prolly/sort/external.go

```diff
@@ -0,0 +1,458 @@
+		return err
+	}
+	if _, err := f.buf.Write(k[:]); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type keyFileReader struct {
+	buf *bufio.Reader
+}
+
+const (
+	keyLenSz = 4
+)
+
+func readUint32(buf []byte) uint32 {
+	return binary.BigEndian.Uint32(buf)
+}
+
+func writeUint32(buf []byte, u uint32) {
+	binary.BigEndian.PutUint32(buf, u)
+}
+
+func (r *keyFileReader) Next(ctx context.Context) (val.Tuple, error) {
+	keySizeBuf := GetKeySizeBuf()
+	defer PutKeySizeBuf(keySizeBuf)
+	if _, err := io.ReadFull(r.buf, keySizeBuf); err != nil {
+		return nil, err
+	}
+
+	keySize := readUint32(keySizeBuf)
+	key := make([]byte, keySize)
+	if _, err := io.ReadFull(r.buf, key); err != nil {
+		return nil, err
+	}
+
+	return key, nil
+}
+
+func (r *keyFileReader) Close() {
+	return
+}
+
+type keyIterable interface {
+	IterAll(context.Context) (keyIter, error)
+}
+
+type keyIter interface {
+	Next(ctx context.Context) (val.Tuple, error)
+	Close()
+}
+
+// mergeFileReader is the heap object for a k-way merge.
+type mergeFileReader struct {
+	// iter abstracts file or in-memory sorted tuples
+	iter keyIter
+	// head is the next tuple in the sorted list
+	head val.Tuple
+}
+
+func (r *mergeFileReader) next(ctx context.Context) (bool, error) {
+	var err error
+	r.head, err = r.iter.Next(ctx)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func newMergeFileReader(ctx context.Context, iter keyIter, heapIdx int) (*mergeFileReader, error) {
+	root, err := iter.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &mergeFileReader{iter: iter, head: root}, nil
+}
+
+type mergeQueue struct {
+	files  []*mergeFileReader
+	keyCmp func(val.Tuple, val.Tuple) bool
+}
+
+func (mq mergeQueue) Len() int { return len(mq.files) }
+
+func (mq mergeQueue) Less(i, j int) bool {
+	// We want Pop to give us the lowest, not highest, priority so we use less than here.
+	return mq.keyCmp(mq.files[i].head, mq.files[j].head)
+}
+
+func (mq mergeQueue) Swap(i, j int) {
+	mq.files[i], mq.files[j] = mq.files[j], mq.files[i]
+}
+
+func (mq *mergeQueue) Push(x any) {
+	item := x.(*mergeFileReader)
+	mq.files = append(mq.files, item)
+}
+
+func (mq *mergeQueue) Pop() any {
+	old := mq.files
+	n := len(old)
+	item := old[n-1]
+	old[n-1] = nil // avoid memory leak
+	mq.files = old[0 : n-1]
+	return item
+}
+
+type fileMerger struct {
+	mq  *mergeQueue
+	out *keyFile
+}
+
+func newFileMerger(ctx context.Context, keyCmp func(val.Tuple, val.Tuple) bool, target *keyFile, files ...keyIterable) (m *fileMerger, err error) {
+	var fileHeads []*mergeFileReader
+	defer func() {
+		if err != nil {
+			for _, fh := range fileHeads {
+				fh.iter.Close()
+			}
+		}
+	}()
+
+	for i, f := range files {
+		iter, err := f.IterAll(ctx)
+		if err != nil {
+			return nil, err
+		}
+		reader, err := newMergeFileReader(ctx, iter, i)
+		if err != nil {
+			iter.Close()
+			if !errors.Is(err, io.EOF) {
```

Comment:
Maybe a comment that`f` was empty and won't make it into the merge queue, but this is not an error.